File: code_samples/section12/lesson_2_linear_search/linear_search.py
"""Linear search with tests."""

import os

#pylint: disable=broad-exception-caught

def linear_search(arr: list[int], target: int) -> int:
    """Basic linear search."""
    for i, value in enumerate(arr):
        if value == target:
            return i
    return -1


def linear_search_steps(arr: list[int], target: int) -> tuple[int, int]:
    """Linear search that returns (index, steps)."""
    steps = 0
    for i, value in enumerate(arr):
        steps += 1
        if value == target:
            return i, steps
    return -1, steps


def load_file(relative_path: str) -> list[int]:
    """Load integers from file relative to script location."""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    full_path = os.path.join(script_dir, relative_path)

    try:
        with open(full_path, "r", encoding="utf-8") as f:
            text = f.read()
        return [int(x) for x in text.split()]
    except Exception as e:
        print(f"Error reading: {full_path}")
        print("Working directory:", os.getcwd())
        print("Exception:", e)
        return []


def run_dataset(name: str, arr: list[int]) -> None:
    """Run a dataset."""
    print(f"=== {name} ===")
    print(f"Loaded {len(arr)} integers\n")

    # First value
    idx, steps = linear_search_steps(arr, arr[0])
    print(f"Search first ({arr[0]}): index={idx}, steps={steps}")

    # Middle value
    mid_value = arr[len(arr) // 2]
    idx, steps = linear_search_steps(arr, mid_value)
    print(f"Search middle ({mid_value}): index={idx}, steps={steps}")

    # Last value
    last_value = arr[-1]
    idx, steps = linear_search_steps(arr, last_value)
    print(f"Search last ({last_value}): index={idx}, steps={steps}")

    # Missing value
    idx, steps = linear_search_steps(arr, 999_999)
    print(f"Search missing (999999): index={idx}, steps={steps}\n")


def main() -> None:
    """Run tests."""
    # Data folder lives at:  ../data/
    ordered   = load_file("../data/ordered.txt")
    unordered = load_file("../data/unordered.txt")

    if not ordered or not unordered:
        print("Missing input files — see error messages above.")
        return

    run_dataset("Ordered Data", ordered)
    run_dataset("Unordered Data", unordered)


if __name__ == "__main__":
    main()

File: code_samples/section12/lesson_2_linear_search/linear_search.c
#include <stdio.h>
#include <stdlib.h>

int linear_search(const int* arr, int n, int target, int* steps_out) {
    int steps = 0;
    for (int i = 0; i < n; i++) {
        steps++;                    // count the comparison arr[i] == target
        if (arr[i] == target) {
            *steps_out = steps;
            return i;
        }
    }
    *steps_out = steps;
    return -1;
}

int* load_file(const char* path, int* out_n) {
    FILE* f = fopen(path, "r");
    if (!f) {
        printf("Error: cannot open %s\n", path);
        *out_n = 0;
        return NULL;
    }

    int capacity = 2048;
    int* arr = malloc(sizeof(int) * capacity);
    int count = 0, x;

    while (fscanf(f, "%d", &x) == 1) {
        if (count == capacity) {
            capacity *= 2;
            arr = realloc(arr, sizeof(int) * capacity);
        }
        arr[count++] = x;
    }

    fclose(f);
    *out_n = count;
    return arr;
}

void run_search_tests(const char* name, const int* arr, int n) {
    printf("=== %s ===\n", name);
    printf("Loaded %d integers.\n\n", n);

    int index, steps;

    // Test 1: First element
    index = linear_search(arr, n, arr[0], &steps);
    printf("Search for first element (%d): index=%d, steps=%d\n",
           arr[0], index, steps);

    // Test 2: Middle element
    index = linear_search(arr, n, arr[n/2], &steps);
    printf("Search for middle element (%d): index=%d, steps=%d\n",
           arr[n/2], index, steps);

    // Test 3: Last element
    index = linear_search(arr, n, arr[n-1], &steps);
    printf("Search for last element (%d): index=%d, steps=%d\n",
           arr[n-1], index, steps);

    // Test 4: Not found case
    index = linear_search(arr, n, 999999, &steps);
    printf("Search for missing element (999999): index=%d, steps=%d\n",
           index, steps);

    printf("\n");
}

int main(void) {
    int n1, n2;

    int* ordered   = load_file("..\\data\\ordered.txt", &n1);
    if (!ordered) return 1;

    int* unordered = load_file("..\\data\\unordered.txt", &n2);
    if (!unordered) {
        free(ordered);
        return 1;
    }

    run_search_tests("Ordered Data", ordered, n1);
    run_search_tests("Unordered Data", unordered, n2);

    free(ordered);
    free(unordered);

    return 0;
}
